From 72cc4511626cca63beb9073c441fb8e6112bd5a0 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Thu, 10 Jan 2013 01:16:39 -0500
Subject: [PATCH] Water flow not trigger chunk loads

Make BlockFromToEvent default cancel if the flow would trigger a chunk load.
---
 src/main/java/org/bukkit/event/block/BlockFromToEvent.java | 9 ++++++++-
 1 file changed, 8 insertions(+), 1 deletion(-)

diff --git a/src/main/java/org/bukkit/event/block/BlockFromToEvent.java b/src/main/java/org/bukkit/event/block/BlockFromToEvent.java
index f976bea..7940b5a 100644
--- a/src/main/java/org/bukkit/event/block/BlockFromToEvent.java
+++ b/src/main/java/org/bukkit/event/block/BlockFromToEvent.java
@@ -21,7 +21,14 @@ public class BlockFromToEvent extends BlockEvent implements Cancellable {
     public BlockFromToEvent(final Block block, final BlockFace face) {
         super(block);
         this.face = face;
-        this.cancel = false;
+        // EMC start- do not flow into unloaded chunks
+        int x = block.getX() + face.getModX();
+        int z = block.getZ() + face.getModZ();
+        this.cancel = !block.getWorld().isChunkLoaded(x >> 4, z >> 4);
+        if (this.cancel) {
+            this.face = BlockFace.SELF;
+        }
+        // EMC end
     }
 
     public BlockFromToEvent(final Block block, final Block toBlock) {
-- 
1.9.1

